src/hdmf/container.py

```diff
@@ -126,6 +128,30 @@ def parent(self, parent_container):
                 parent_container.__children.append(self)
                 parent_container.set_modified()
 
+    def __deepcopy__(self, memo):
+        ''' Create a deep copy of this Container
+        Reset the root parent to None, set all container_source to None, and set new object_id value
+        '''
+        deepcopy_method = self.__deepcopy__
+        self.__deepcopy__ = None
+        cp = deepcopy(self, memo)
+        self.__deepcopy__ = deepcopy_method
+
+        cp.__parent = None
+        cp.__container_source = None
+        cp.__object_id = str(uuid4())
+        cp.__modified = True
+
+        for child in cp.children:
+            child.parent = cp
+
+        # resolve HDMFDataset after and separately because DataIO can wrap an HDMFDataset
+        for k, v in cp.__dict__.items():
+            if isinstance(v, HDMFDataset):
+                setattr(cp, k, v.dataset)
```

Comment:
On read ``v.dataset`` may be and ``h5py.Dataset`` with references. I think the resolution for this would be something like

```
if len(v) > 0 and isinstance(v[0], Container): 
    setattr(cp, k, [find_copied_container(c) for c in v])
else:
    setattr(cp, k, v.dataset)
```
where ``find_copied_container`` would need to replace the reference with the new copy. Not sure, what would happen if you tried to call deepcopy on, e.g., a single DynamicTable that contains references, as in this case those containers may not have been copied yet by deepcopy. 

Also, I'm not sure what happens when the container that is being copied was loaded from a file. In this case you may have h5py.Dataset objects that would need to be replaced as well (or are those all wrapped?).